### ISSUE TITLE
Management command returns code 0 even if an error occured

### ISSUE THREAD

**Issue by gwaldvogel** [translate/pootle#4690]
I am currently trying to automate loading the translations into Pootle via Jenkins and came across an error with the update_stores management command.

Even though there is an error (see below) the return code is 0 which makes it hard to check whether or not the update worked as intended.

```pytb
2016-04-28 14:58:02,901 ERROR Failed to run update_stores over /ru/xxxxxxxx/ru_RU.ts
Traceback (most recent call last):
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_app/management/commands/__init__.py", line 71, in do_translation_project
    self.handle_store(store, **options)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_app/management/commands/update_stores.py", line 50, in handle_store
    store.update(overwrite=overwrite, only_newer=not force)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1974, in update
    submission_type=submission_type)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1699, in update_units
    current_time, user, submission_type)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1767, in record_submissions
    sub.save()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_statistics/models.py", line 357, in save
    super(Submission, self).save(*args, **kwargs)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 589, in save
    force_update=force_update, update_fields=update_fields)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 617, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 698, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 731, in _do_insert
    using=using, raw=raw)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 921, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 921, in execute_sql
    cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 129, in execute
    return self.cursor.execute(query, args)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 187, in execute
    query = query % tuple([db.literal(item) for item in args])
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 278, in literal
    return self.escape(o, self.encoders)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 203, in string_literal
    return db.string_literal(obj)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/translate/misc/autoencode.py", line 64, in __str__
    return super(autoencode, self).__str__()
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-8: ordinal not in range(128)
```

**Comment by unho** [comment]
@gwaldvogel Which Pootle version are you using? Which Translate Toolkit version do you have installed?

This error looks like some elusive error we managed to track down and fix in Translate Toolkit master: https://github.com/translate/pootle/issues/4654

**Comment by gwaldvogel** [comment]
I installed Pootle and the Translate Toolkit via pip with the versions 1.13.0 (Translate Toolkit) and 2.7.2 (Pootle).
The error above is because of a broken .ts XML schema (there is a <numerusform> although the source string does not contain a %n).

The problem  I am reporting is not that an error occurs, but that the command returns code 0 which means everything is alright, although the import did not work.

**Comment by phlax** [comment]
@gwaldvogel thanks for the feedback

i think that the problem is that the error is not detected/handled, and so the command returns without failing.

We need to know which version of pootle/ttk you are using in order to try and reproduce the problem, and to see if its still present in the latest versions.

If you were able to provide a copy of the file that triggers that would be huge help in saving our time.

If you want to pm any of us the file you can usually find us on https://gitter.im/translate/pootle

**Comment by gwaldvogel** [comment]
@phlax thanks a lot for your fast reply. Like I already mentioned, I am using Pootle version 2.7.2 and Translate Toolkit 1.13.0.
```
$ pootle --version
Pootle 2.7.2 (Django 1.7.11, Translate Toolkit 1.13.0)
```
I will prepare a demo file and send it to you.

**Comment by gwaldvogel** [comment]
@phlax You can use the following code to reproduce the error:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="de_DE" sourcelanguage="en_GB">
<context>
    <name>Attestation</name>
    <message numerus="yes">
        <source>Attachment</source>
        <translation>
            <numerusform>Anhang</numerusform>
            <numerusform>Anhänge</numerusform>
        </translation>
    </message>
</context>
</TS>
```
(from the German translation file)
and
```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en_GB" sourcelanguage="en_GB">
<context>
    <name>Attestation</name>
    <message numerus="yes">
        <source>Attachment</source>
        <translation>
            <numerusform>Attachment</numerusform>
            <numerusform>Attachments</numerusform>
        </translation>
    </message>
</context>
</TS>
```
(from the English translation file)

I just tried it with Pootle version 2.7.3, the command now correctly responds with error code 1, but the plural is not properly handled.

**Comment by phlax** [comment]
@gwaldvogel im off for a few days but will follow up when i get chance. I have a feeling that this is fixed in master/upcoming 2.8.0 with ttk2.0 but id need to check

**Comment by ta2-1** [comment]
Hi @gwaldvogel. @phlax asked me to take this issue.
Could you elaborate what you meant with "but the plural is not properly handled"?

**Comment by phlax** [comment]
@ta2-1 i think we just need to test with the sample file using old ttk/pootle to trigger error, and then check again with master and see that its gone (hopefully)

**Comment by gwaldvogel** [comment]
@ta2-1 Usually you have a placeholder within the text, where the amount is inserted, like "%n attachments" which would be translated to "1 attachment" or "2 attachments", but in my example we do not display the actual amount. The TTK did not correctly recognize that this is a correct syntax, because there is no %n within the source string. This plural form is handled within the .ts files by the `numerus="yes"` and the different `<numerusform>`'s.

**Comment by ta2-1** [comment]
@gwaldvogel thanks for your explanation.
I'd like to add test for the initial issue and I tried to reproduce this error
```
File "/var/www/pootle/env/local/lib/python2.7/site-packages/translate/misc/autoencode.py", line 64, in __str__
    return super(autoencode, self).__str__()
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-8: ordinal not in range(128)
``` 
which comes from ttk 1.13.0. But I couldn't do it.
What I did:
- set up Pootle 2.7.2 and ttk 1.13.0;
- added a test project;
- created directories for languages `en` and `de` in filesystem;
- created files with content you provided;
- used `update_stores --project=test` to fill the project;
- deleted translations from `de` units via web UI;
- run `update_stores --project=test --language=de`;
I could see no errors and plurals worked properly i.e. translations were added.

As I see the initial error was caused by `translate/misc/autoencode.py` which was already removed. I'd say that this issue was fixed. But you mentioned that 
> the command now correctly responds with error code 1

So my question is: are you able to reproduce this bug on 2.7.3? Is traceback the same?

I can confirm that plurals don't work properly on later Pootle and ttk versions. I'm checking what causes that problem. You mentioned that
> The TTK did not correctly recognize that this is a correct syntax.

TTK worked properly in my tests and I suspect that problem is when we update (save to DB) plurals in Pootle.
So I'll provide more information here when I have it.

**Comment by gwaldvogel** [comment]
@ta2-1 this is the traceback I had with Pootle 2.7.3 and TTK 1.13.0.
It occurs right after the singular form is shown in the log:

```
[2016-04-29T14:06:06]	system	UA	be	268349	/be/testproject/fr_BE.ts	Pièce jointe
Traceback (most recent call last):
  File "/var/www/pootle/env/bin/pootle", line 11, in <module>
    sys.exit(main())
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/runner.py", line 319, in main
    django_settings_module='pootle.settings')
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/runner.py", line 299, in run_app
    management.execute_from_command_line(command)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 385, in execute_from_command_line
    utility.execute()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 377, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 338, in execute
    output = self.handle(*args, **options)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 533, in handle
    return self.handle_noargs(**options)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_app/management/commands/__init__.py", line 119, in handle_noargs
    self.handle_all(**options)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_app/management/commands/update_stores.py", line 67, in handle_all
    projects=self.projects)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_translationproject/models.py", line 89, in scan_translation_projects
    create_or_resurrect_translation_project(language, project)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_translationproject/models.py", line 46, in create_or_resurrect_translation_project
    tp = create_translation_project(language, project)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_translationproject/models.py", line 61, in create_translation_project
    .get_or_create(language=language, project=project)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 424, in get_or_create
    return self._create_object_from_params(lookup, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 457, in _create_object_from_params
    obj.save(force_insert=True, using=self.db)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_translationproject/models.py", line 302, in save
    changed = store.update_from_disk()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1833, in update_from_disk
    store_revision=store_revision,
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1777, in update
    self.save(update_cache=has_changed)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 1460, in save
    super(Store, self).save(*args, **kwargs)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 589, in save
    force_update=force_update, update_fields=update_fields)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 617, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 679, in _save_table
    forced_update)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 723, in _do_update
    return filtered._update(values) > 0
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 600, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 1005, in execute_sql
    cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 787, in execute_sql
    cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 59, in execute
    self.db.validate_no_broken_transaction()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/__init__.py", line 386, in validate_no_broken_transaction
    "An error occurred in the current transaction. You can't "
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

**Comment by ta2-1** [comment]
@gwaldvogel known plurals bug was fixed in #4707. 
I think that this one from your traceback is different. But I can't reproduce it with examples you provided.
Can we close this bug where title is unrelated already and open new one?

@phlax I wonder if this last traceback can be related to having different plural forms for a different unit with the same source string (which is unsupported by ttk as I understood)?

**Comment by gwaldvogel** [comment]
@ta2-1 I opened issue #4727 for that.
